src/main/java/org/folio/entlinks/service/authority/AuthoritySourceFileService.java

```diff
@@ -0,0 +1,109 @@
+import org.folio.entlinks.domain.entity.AuthoritySourceFile;
+import org.folio.entlinks.domain.repository.AuthoritySourceFileRepository;
+import org.folio.entlinks.domain.repository.AuthorityStorageRepository;
+import org.folio.entlinks.exception.AuthoritySourceFileNotFoundException;
+import org.folio.entlinks.exception.RequestBodyValidationException;
+import org.folio.spring.data.OffsetRequest;
+import org.folio.tenant.domain.dto.Parameter;
+import org.springframework.data.domain.Page;
+import org.springframework.stereotype.Service;
+
+@Service
+@AllArgsConstructor
+@Log4j2
+public class AuthoritySourceFileService {
+
+  private final AuthoritySourceFileRepository repository;
+  private final AuthoritySourceFileMapper mapper;
+  private final AuthorityStorageRepository authorityStorageRepository;
+
+  public Page<AuthoritySourceFile> getAll(Integer offset, Integer limit, String cql) {
+    log.debug("getAll:: Attempts to find all AuthoritySourceFile by [offset: {}, limit: {}, cql: {}]", offset, limit,
+      cql);
+
+    if (StringUtils.isBlank(cql)) {
+      return repository.findAll(new OffsetRequest(offset, limit));
+    }
+
+    return repository.findByCql(cql, new OffsetRequest(offset, limit));
+  }
+
+  public AuthoritySourceFile getById(UUID id) {
+    log.debug("getById:: Loading AuthoritySourceFile by ID [id: {}]", id);
+
+    return repository.findById(id).orElseThrow(() -> new AuthoritySourceFileNotFoundException(id));
+  }
+
+  public AuthoritySourceFile create(AuthoritySourceFile entity) {
+    log.debug("create:: Attempting to create AuthoritySourceFile [entity: {}]", entity);
+
+    UUID id = entity.getId();
+    if (id != null && repository.existsById(id)) {
+      throw new RequestBodyValidationException("AuthoritySourceFile already exists",
+        List.of(new Parameter("id").value(String.valueOf(id))));
+    }
+
+    for (var code : entity.getAuthoritySourceFileCodes()) {
+      code.setAuthoritySourceFile(entity);
+    }
+
+    return repository.save(entity);
+  }
+
+  public AuthoritySourceFile update(UUID id, AuthoritySourceFile modified) {
+    log.debug("update:: Attempting to update AuthoritySourceFile [id: {}]", id);
+
+    if (!Objects.equals(id, modified.getId())) {
+      throw new RequestBodyValidationException("Request should have id = " + id,
+        List.of(new Parameter("id").value(String.valueOf(modified.getId()))));
+    }
+
+    var existingEntity = repository.findById(id).orElseThrow(() -> new AuthoritySourceFileNotFoundException(id));
+
+    copyModifiableFields(existingEntity, modified);
+
+    return repository.save(existingEntity);
+  }
+
+  public void deleteById(UUID id) {
+    log.debug("deleteById:: Attempt to delete AuthoritySourceFile by [id: {}]", id);
+
+    var sourceFile = repository.findById(id)
+        .orElseThrow(() -> new AuthoritySourceFileNotFoundException(id));
+
+    authorityStorageRepository.findBySourceFileId(id).forEach(sourceFile::removeAuthorityStorage);
```

Comment:
We have to restrict deleting of the ASF if it has references in authorities. In this case we can rely on database constraints and process it in APIErrorHandler